File: engine/src/main/java/de/hanno/hpengine/engine/graphics/renderer/PixelBufferObject.java
package de.hanno.hpengine.engine.graphics.renderer;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL15.*;
import static org.lwjgl.opengl.GL21.*;

import java.nio.ByteBuffer;

import org.lwjgl.BufferUtils;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.GL13;
import org.lwjgl.opengl.GL30;
import de.hanno.hpengine.engine.graphics.renderer.constants.GlTextureTarget;

public class PixelBufferObject {
	
	private int id;
	private int width;
	private int height;
	private ByteBuffer buffer;
	private float[] array;

	public PixelBufferObject(int width, int height) {
		id = glGenBuffers();
		this.width = width;
		this.height = height;
		this.buffer = BufferUtils.createByteBuffer(4*4*width*height); // 4 is byte size of float
		this.array = new float[4*width*height];
		bind();
		glBufferData(GL_PIXEL_UNPACK_BUFFER, 4*4*width*height, GL_DYNAMIC_READ);
		unbind();
	}

	private void unbind() {
		glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0);
	}
	
	public void bind() {
		glBindBuffer(GL_PIXEL_UNPACK_BUFFER, id);
	}

	public void readPixelsFromTexture(int textureId, int mipmapLevel, GlTextureTarget target, int format, int type) {
		bind();
        GraphicsContext.getInstance().bindTexture(target, textureId);
		glGetTexImage(target.glTarget, mipmapLevel, format, type, buffer);
		unbind();
	}
	public void glTexSubImage2D(int textureId, int mipmapLevel, GlTextureTarget target, int format, int type, int width, int height, ByteBuffer buffer) {
		glTexSubImage2D(textureId, mipmapLevel, target, format, type, 0, 0, width, height, buffer);
	}
	public void glTexSubImage2D(int textureId, int mipmapLevel, GlTextureTarget target, int format, int type, int offsetX, int offsetY, int width, int height, ByteBuffer buffer) {
		mapAndUnmap(offsetX, offsetY, width, height, buffer);
        GraphicsContext.getInstance().execute(() -> {
            GraphicsContext.getInstance().bindTexture(target, textureId);
			GL11.glTexSubImage2D(target.glTarget, mipmapLevel, offsetX, offsetY, width, height, GL_RGBA, GL_FLOAT, 0);
		});
		unbind();
	}

	public void glCompressedTexImage2D(int textureId, GlTextureTarget target, int level, int internalformat, int width, int height, int border, ByteBuffer textureBuffer) {
        GraphicsContext.getInstance().execute(() -> {
			mapAndUnmap(0, 0, width, height, buffer);
            GraphicsContext.getInstance().bindTexture(target, textureId);
			GL13.glCompressedTexImage2D(target.glTarget, level, internalformat, width, height, border, 0);
		});
		unbind();
	}

	private void mapAndUnmap(int offsetX, int offsetY, int width, int height, ByteBuffer buffer) {
		bind();
//		glBufferData(GL_PIXEL_UNPACK_BUFFER, 4*4*(width-offsetX)*(height*offsetY), GL_STREAM_COPY);
//		ByteBuffer result = glMapBuffer(GL_PIXEL_UNPACK_BUFFER, GL_READ_WRITE, buffer);
		ByteBuffer result = GL30.glMapBufferRange(GL_PIXEL_UNPACK_BUFFER, 0, 4*4*(width)*(height), GL30.GL_MAP_READ_BIT, buffer);
		result.put(buffer);
		glUnmapBuffer(GL_PIXEL_UNPACK_BUFFER);
	}

	public float[] mapBuffer() {
		glMapBuffer(GL_PIXEL_UNPACK_BUFFER, GL_READ_WRITE, buffer);
		buffer.rewind();
		buffer.asFloatBuffer().get(array);
		return array;
	}
}

File: opengl/src/main/kotlin/de/hanno/hpengine/graphics/renderer/PixelBufferObject.kt
package de.hanno.hpengine.graphics.renderer

import de.hanno.hpengine.graphics.GraphicsApi
import de.hanno.hpengine.graphics.renderer.constants.TextureTarget
import de.hanno.hpengine.graphics.renderer.constants.glValue
import org.lwjgl.BufferUtils
import org.lwjgl.opengl.*
import java.nio.ByteBuffer

class PixelBufferObject(private val graphicsApi: GraphicsApi, private val width: Int, private val height: Int) {
    private val id: Int = GL15.glGenBuffers()
    private val buffer: ByteBuffer = BufferUtils.createByteBuffer(4 * 4 * width * height) // 4 is byte size of float
    private val array: FloatArray = FloatArray(4 * width * height)

    init {
        bind()
        GL15.glBufferData(GL21.GL_PIXEL_UNPACK_BUFFER, (4 * 4 * width * height).toLong(), GL15.GL_DYNAMIC_READ)
        unbind()
    }

    fun unbind() {
        GL15.glBindBuffer(GL21.GL_PIXEL_UNPACK_BUFFER, 0)
    }

    fun bind() {
        GL15.glBindBuffer(GL21.GL_PIXEL_UNPACK_BUFFER, id)
    }

    fun readPixelsFromTexture(textureId: Int, mipmapLevel: Int, target: TextureTarget, format: Int, type: Int) {
        bind()
        graphicsApi.bindTexture(target, textureId)
        GL11.glGetTexImage(target.glValue, mipmapLevel, format, type, buffer)
        unbind()
    }

    fun glTexSubImage2D(
        textureId: Int,
        mipmapLevel: Int,
        target: TextureTarget,
        format: Int,
        type: Int,
        width: Int,
        height: Int,
        buffer: ByteBuffer?
    ) {
        glTexSubImage2D(textureId, mipmapLevel, target, format, type, 0, 0, width, height, buffer)
    }

    fun glTexSubImage2D(
        textureId: Int,
        mipmapLevel: Int,
        target: TextureTarget,
        format: Int,
        type: Int,
        offsetX: Int,
        offsetY: Int,
        width: Int,
        height: Int,
        buffer: ByteBuffer?
    ) {
        mapAndUnmap(offsetX, offsetY, width, height, buffer)
        graphicsApi.onGpu {
            graphicsApi.bindTexture(target, textureId)
            GL11.glTexSubImage2D(
                target.glValue,
                mipmapLevel,
                offsetX,
                offsetY,
                width,
                height,
                GL11.GL_RGBA,
                GL11.GL_FLOAT,
                0
            )
            Unit
        }
        unbind()
    }

    fun glCompressedTexImage2D(
        textureId: Int,
        target: TextureTarget,
        level: Int,
        internalformat: Int,
        width: Int,
        height: Int,
        border: Int,
        textureBuffer: ByteBuffer?
    ) {
        graphicsApi.onGpu {
            mapAndUnmap(0, 0, width, height, buffer)
            graphicsApi.bindTexture(target, textureId)
            GL13.glCompressedTexImage2D(target.glValue, level, internalformat, width, height, border, null)
            Unit
        }
        unbind()
    }

    private fun mapAndUnmap(offsetX: Int, offsetY: Int, width: Int, height: Int, buffer: ByteBuffer?) {
        bind()
        //		glBufferData(GL_PIXEL_UNPACK_BUFFER, 4*4*(width-offsetX)*(height*offsetY), GL_STREAM_COPY);
//		ByteBuffer result = glMapBuffer(GL_PIXEL_UNPACK_BUFFER, GL_READ_WRITE, buffer);
        val result = GL30.glMapBufferRange(
            GL21.GL_PIXEL_UNPACK_BUFFER,
            0,
            (4 * 4 * width * height).toLong(),
            GL30.GL_MAP_READ_BIT,
            buffer
        )
        result!!.put(buffer)
        if (GL15.glGetBufferParameteri(GL21.GL_PIXEL_UNPACK_BUFFER, GL15.GL_BUFFER_MAPPED) == GL11.GL_TRUE) {
            GL15.glUnmapBuffer(GL21.GL_PIXEL_UNPACK_BUFFER)
        }
    }

    fun mapBuffer(): FloatArray {
        GL15.glMapBuffer(GL21.GL_PIXEL_UNPACK_BUFFER, GL15.GL_READ_WRITE, buffer)
        buffer.rewind()
        buffer.asFloatBuffer()[array]
        return array
    }
}